### PR TITLE
update php/object tag syntax

### LINKED ISSUE #16877: YAML PHP object tag property specification is incorrect
I came to notice that the tag property for php objects in the YAML parser/dumper is `!!php/object` which, according to the spec, means `tag:yaml.org,2002:php/object`, which in turn is incorrect because `php` isn't a actual type represented by YAML.

So I think it should be a local `!php/object` (single exclamation, local tag), which conveniently enough matches with the [php-yaml extension](https://secure.php.net/manual/en/ref.yaml.php) (LibYAML wrapper).

Excerpt from ref: http://yaml.org/refcard.html
```
Tag property: # Usually unspecified.
    none    : Unspecified tag (automatically resolved by application).
    '!'     : Non-specific tag (by default, "!!map"/"!!seq"/"!!str").
    '!foo'  : Primary (by convention, means a local "!foo" tag).
    '!!foo' : Secondary (by convention, means "tag:yaml.org,2002:foo").
    '!h!foo': Requires "%TAG !h! <prefix>" (and then means "<prefix>foo").
    '!<foo>': Verbatim tag (always means "foo").
```

### PR DESCRIPTION
**Pull request by koenpunt**
[Yaml] deprecate old php/object syntax

| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #16877 |
| License | MIT |

the old syntax (double exclamation mark) is to represent native YAML
types, a single exclamation mark represent a local tag.
an E_USER_DEPRECATED error is triggered for old tags.

### CODE REVIEW

**Comment by xabbuh** [comment]
Actually, I would consider this a bugfix which means that the changes here (except for the deprecation trigger) should be done in 2.3.

File: src/Symfony/Component/Yaml/Inline.php [hunk 1]
```diff
@@ -477,8 +477,11 @@ private static function evaluateScalar($scalar, $references = array())
                     case 0 === strpos($scalar, '! '):
                         return (int) self::parseScalar(substr($scalar, 2));
                     case 0 === strpos($scalar, '!!php/object:'):
+                        trigger_error('!!php/object is deprecated use !php/object instead', E_USER_DEPRECATED);
+                        $scalar = substr($scalar, 1);
```
Comment (stof) [inline src/Symfony/Component/Yaml/Inline.php]: I would add a comment saying the fallthrough is intended

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 2]
```diff
@@ -443,7 +443,29 @@ public function testObjectSupportDisabledButNoExceptions()
      */
     public function testObjectsSupportDisabledWithExceptions()
     {
-        $this->parser->parse('foo: !!php/object:O:30:"Symfony\Tests\Component\Yaml\B":1:{s:1:"b";s:3:"foo";}', true, false);
+        $this->parser->parse('foo: !php/object:O:30:"Symfony\Tests\Component\Yaml\B":1:{s:1:"b";s:3:"foo";}', true, false);
+    }
+
+    /**
+     * @expectedException PHPUnit_Framework_Exception
+     * @expectedExceptionMessage !!php/object is deprecated use !php/object instead
```
Comment (stof) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: this is wrong. There is no exception with such message

File: src/Symfony/Component/Yaml/Tests/ParserTest.php [hunk 3]
```diff
@@ -443,7 +443,29 @@ public function testObjectSupportDisabledButNoExceptions()
...
+    }
+
+    public function testDeprecatedObjectNotationWithErrorsSuppressed()
+    {
+        $input = <<<EOF
+foo: !!php/object:O:30:"Symfony\Component\Yaml\Tests\B":1:{s:1:"b";s:3:"foo";}
+bar: 1
+EOF;
+
+        $previous = error_reporting(E_ALL & ~E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: this is *not* how we handle deprecations (because you miss the silencing)

**Comment by stof** [comment]
I agree with @xabbuh that we should consider the proper syntax as a bugfix. The deprecation should then be added in master only

File: src/Symfony/Component/Yaml/Inline.php [hunk 4]
```diff
@@ -477,8 +477,11 @@ private static function evaluateScalar($scalar, $references = array())
                     case 0 === strpos($scalar, '! '):
                         return (int) self::parseScalar(substr($scalar, 2));
                     case 0 === strpos($scalar, '!!php/object:'):
+                        trigger_error('!!php/object is deprecated use !php/object instead', E_USER_DEPRECATED);
```
Comment (stof) [inline src/Symfony/Component/Yaml/Inline.php]: missing silencing (which makes the testsuite fail btw)

File: src/Symfony/Component/Yaml/Tests/ParserTest.php (hunk 2 above)
Comment (koenpunt) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: PHPUnit converts errors to exceptions, so this exception actually is thrown.
 Or isn't that what you mean?

File: src/Symfony/Component/Yaml/Tests/ParserTest.php (hunk 3 above)
Comment (koenpunt) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: Then how do we?

File: src/Symfony/Component/Yaml/Tests/ParserTest.php (hunk 2 above)
Comment (stof) [inline src/Symfony/Component/Yaml/Tests/ParserTest.php]: http://symfony.com/doc/current/contributing/code/conventions.html#deprecations

**Comment by koenpunt** [comment]
Oh well, there's a difference in running the Yaml suite and running the full suite...

**Comment by xabbuh** [comment]
@koenpunt Can you only add the deprecation here and create a new pull request for the bugfix based on the `2.3` branch?

File: src/Symfony/Component/Yaml/Inline.php [hunk 5]
```diff
@@ -477,8 +477,11 @@ private static function evaluateScalar($scalar, $references = array())
                     case 0 === strpos($scalar, '! '):
                         return (int) self::parseScalar(substr($scalar, 2));
                     case 0 === strpos($scalar, '!!php/object:'):
+                        @trigger_error('!!php/object is deprecated since version 3.0 and will be removed in 4.0. Use !php/object instead.', E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Component/Yaml/Inline.php]: The version this will be deprecated in is 3.1 (not 3.0).

**Comment by xabbuh** [comment]
closing in favour of #17461 and #17462

**Comment by koenpunt** [comment]
:ok_hand:

**Pull request closed without merging**